### PR TITLE
RATIS-1874 Add notifyLeaderReady function in IStateMachine

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**
see [jira](https://issues.apache.org/jira/browse/RATIS-1874)

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 1]
```diff
@@ -1825,6 +1828,11 @@ CompletableFuture<Message> applyLogToStateMachine(LogEntryProto next) throws Raf
     return null;
   }
 
+  boolean isFirstLogAppliedByCurrentLeaderInCurrentTerm(long term) {
+    return term != state.getLastEntry().getTerm()
```
Comment (SzyWilliam) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:1832]: Thanks for the patch! I guess you are trying to compare the term of the _current applying log_ with the _last applied log_ to decide the very first log of the current leader. However, `state.getLastEntry().getTerm()` returns the last **committed** log's term, not the last **applied** log's term.

Comment (OneSizeFitsQuorum) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java]: Yes, I made a mistake. Has fixed~

**Comment by szetszwo** [comment]
Hi @OneSizeFitsQuorum , In Apache Ozone, we detect leader ready by the following steps

1. Record the leader term in `notifyLeaderChanged`.
https://github.com/apache/ozone/blob/fcf5b17a4ede0c55c76aa337438498361c0a5dd3/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java#L282-L283

2. Compare the term in `notifyTermIndexUpdated` and check `DivisionInfo.isLeaderReady()`.
https://github.com/apache/ozone/blob/fcf5b17a4ede0c55c76aa337438498361c0a5dd3/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMStateMachine.java#L357-L361

Please take a look if it is good for your use case.

**Comment by OneSizeFitsQuorum** [comment]
Hi, @szetszwo~ In theory a similar approach could be taken to achieve the same purpose, but we are also wondering if there is a more direct scheme, such as this interface we have added.

Do you think it is necessary to add such a more direct interface?

**Review by szetszwo — COMMENTED** [review]
@OneSizeFitsQuorum , sure, let's do it.

We can also improve the `isReady()` method; see https://issues.apache.org/jira/secure/attachment/13062432/906_review.patch

File: ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java [hunk 2]
```diff
@@ -150,6 +150,13 @@ interface EventApi {
      */
     default void notifyLeaderChanged(RaftGroupMemberId groupMemberId, RaftPeerId newLeaderId) {}
 
+    /**
+     * Notify the {@link StateMachine} that current node has become the leader and recovered successfully.
+     * Note that only leader will call this function.
+     *
+     */
+    default void notifyLeaderReady() {}
```
[the remark this answers is not in the document]
Comment (OneSizeFitsQuorum) [reply ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java]: A more reasonable idea!

**Comment by OneSizeFitsQuorum** [comment]
@szetszwo, thank you for the more maintainable Patch!

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

File: ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java [hunk 3]
```diff
@@ -228,6 +228,11 @@ default void notifyFollowerSlowness(RoleInfoProto leaderInfo) {}
      * Notify {@link StateMachine} that this server is no longer the leader.
      */
     default void notifyNotLeader(Collection<TransactionContext> pendingEntries) throws IOException {}
+
+    /**
+     * Notify the {@link StateMachine} that this server becomes ready after changed to leader.
+     */
+    default void notifyLeaderReady() {}
```
Comment (tisonkun) [inline ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java:235]: This method name can be a bit misleading -

* What I read: a leader is ready for this group.
* What it actually means: this server is reader after changed to leader.

Comment (tisonkun) [reply ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java:235]: Perhaps `notifyServingAsLeader` or `notifyBecomeLeader`

Comment (OneSizeFitsQuorum) [reply ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java:235]: > What it actually means: this server is reader after changed to leader.

Is `reader` a typo for `ready`?

This function means that the current server has become a leader and has fully recovered the state machine (at least one log entry from the current term has been applied).

If we use the name "notifyBecomeLeader," it seems to suggest that the current peer has been elected as a leader but doesn't explicitly convey the semantics of the state machine recovery. "notifyServingAsLeader" and "notifyLeaderReady" both imply some level of state machine recovery in my view, and either of them could work for me.

What's your opinion? @szetszwo @tisonkun @SzyWilliam

Comment (tisonkun) [reply ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java:235]: > the javadoc is clear

This is a good point. The method name is conveyed by `LeaderStateImpl#isReady` with its doc clear. So `notifyLeaderReady` should be a good name follow this idiom >_<

**Pull request merged**